src/main/java/com/amazon/corretto/hotpatch/HotPatchAgent.java

```diff
@@ -15,107 +15,321 @@
+ *
+ * What patcher to install and the location of the jar where the patcher is present (often bundled in the same one as
+ * the agent) are passed as parameters to the agent. If the agent is not able to locate or load the jar using its
+ * custom classloader, the SystemClassloader will be used, mirroring the behavior of previous versions of the tool. If
+ * the patcher to be loaded cannot be determined, the default patcher
+ * {@link com.amazon.corretto.hotpatch.patch.impl.set.Log4j2PatchSetV1} will be applied, which applies the same
+ * class transformation as older versions of this tool.
+ */
 public class HotPatchAgent {
+  // version of this agent, it will be stored in a property named {@link Constants#LOG4J_FIXER_AGENT_VERSION}
+  public static final int HOTPATCH_AGENT_VERSION = 2;
 
-  // version of this agent
-  private static final int log4jFixerAgentVersion = 1;
-
+  // used to know if we are the first time the agent is being loaded
   private static boolean agentLoaded = false;
-  private static boolean staticAgent = false; // Set to true if loaded as a static agent from 'premain()'
 
-  public static void premain(String args, Instrumentation inst) {
-    staticAgent = true;
-    agentmain(args, inst);
+  // a default implementation of the logger. By default, set to verbose, but information regarding verbosity will be set
+  // before any message is executed. Subsequent installations of the agent will not create new loggers, but they will
+  // update it to match the new configuration.
+  private static final LoggerImpl logger = new LoggerImpl();
+  private static final Map<String, Patcher> appliedPatchers = new HashMap<>();
+
+  // Some patchers may load ASM from their own jar, others will use the version already loaded if an agent is already
+  // present. We pick the version of ASM included with the first agent attached as the one to use.
+  public static int asmApiVersion() {
+    return Opcodes.ASM9;
   }
 
-  private static List<HotPatch> loadPatches(String args) {
-    List<HotPatch> patches = new ArrayList<>();
-    if (Log4j2NoJndiLookup.isEnabled(args)) {
-      patches.add(new Log4j2NoJndiLookup());
+  /**
+   * This is the entry point when the agent is loaded during startup. The main difference in this scenario will be that
+   * we only need to load our transformers, but there is no need to retransform existing classes as they have not been
+   * loaded yet. Additionally, we may not receive parameters as agent args, but as SystemProperties.
+   * @param args String representing the different arguments for the agent.
+   * @param inst An instance of instrumentation that will be used to apply the different patchers.
+   */
+  public static void premain(String args, Instrumentation inst) {
+    if (args == null || !args.contains(PATCHER_NAME_ARG)) {
+      // The default patcher class. Change this if you add a new version of the Patcher.
+      String userPatcherClass = System.getProperty(PATCHER_NAME_ARG);
```

Comment:
I know this code is coming from my patch, but now, after thinking about it, I don't think we need to query the system properties for the static agent case. If used as a static agent, we can simply pass the arguments we need directly to the agent.
I think the property is only required when calling the command line tool to simplify argument handling.